packages/web/src/components/DashboardPanelCard.tsx

```
@@ -167,14 +259,31 @@ export default function DashboardPanelCard({
-    [panel.queries, panel.query, panel.visualization]
+          : [];
+      // Substitute `${datasource}` (or any `${name}`) placeholders in
+      // datasourceId so a `$datasource` template variable can swap backends
+      // at view time without rewriting the panel.
+      if (!variableValues) return baseQueries;
+      return baseQueries.map((q) => ({
+        ...q,
+        datasourceId: substituteVariableTokens(q.datasourceId, variableValues),
+      }));
+    },
+    [panel.queries, panel.query, panel.visualization, variableValues, variableValuesKey] // eslint-disable-line react-hooks/exhaustive-deps
```

ESLint

[error] Definition for rule 'react-hooks/exhaustive-deps' was not found.

(react-hooks/exhaustive-deps)